### PR TITLE
Improve dictionary with dynamic query entries 

### PR DESCRIPTION
**Pull request by jiqiang90**
# Description

There is not much improvement we could make for query distinct. 

However, this pr improves query entries. Rather than use queryEndBlock, we dynamic monitor an attempting query response, if the endBlock is beyond a new datasource, we will update the query entries, and resend the query again.

It will save some queries and block when fetch in between startHeight to queryEndblock. 

- [ ] Haven't check if multiple ds with same startHeight could generate incorrect map?

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] I have tested locally
- [ ] I have performed a self review of my changes
- [ ] Updated any relevant documentation
- [ ] Linked to any relevant issues
- [ ] I have added tests relevant to my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] My code is up to date with the base branch

### CODE REVIEW

**Review by stwiname — CHANGES_REQUESTED** [review]
I don't think we're gaining anything with this.

We're making dictionary queries and discarding the results. We can be better about setting the queryEndBlock [here](https://github.com/subquery/subql/blob/585ceb34ea721f7c83dec71bf9d82d0f0f98cfc8/packages/node/src/indexer/fetch.service.ts#L431) and consider the keys of [mappedDictionaryQueryEntries](https://github.com/subquery/subql/blob/585ceb34ea721f7c83dec71bf9d82d0f0f98cfc8/packages/node-core/src/indexer/dictionary.service.ts#L138) to decide on the end value.

File: packages/node-core/src/indexer/dictionary.service.ts [hunk 1]
```diff
@@ -198,6 +203,11 @@ export class DictionaryService implements OnApplicationShutdown {
...
       const _metadata = resp.data._metadata;
       const endBlock = Math.min(...Object.values(entityEndBlock).map((height) => (isNaN(height) ? Infinity : height)));
+
+      // This filter with endBlock, will ensure next query for DS is not lost
+      // Example, Ds1 return 7,8,9, Ds2 return 15,16,17
+      // Batch blocks will only be 7,8,9
+      // next query start is 10, not 17. So between 10-17 for Ds1 will not missing.
       const batchBlocks = Array.from(blockHeightSet)
         .filter((block) => block <= endBlock)
         .sort((n1, n2) => n1 - n2);
```
[the remark this answers is not in the document]
Comment (jiqiang90) [reply packages/node-core/src/indexer/dictionary.service.ts:213]: Setting to right end block will lead some blocks missing for other filters/conditions. 
Maybe the comment here is not clear enough, let me try again:

Assume dictionary we have to query both events and extrinsics:

Query conditions:
startBlock: 10, queryEndBlock: 210,  first: 30

Query for `staking.Rewards` **events** returns:
[10, 12, 15 ...  50].     total 30 blocks, end with block height 50

Query for `balances.Transfer` **extrinsics** returns:
[15, 19, 25 ...  150].     total 30 blocks, end with block height 150


And if we directly join these two batch blocks together, set endBlock to 150, next dictionary query will start with block height 151.

In that case, we are missing from block 51 -> 150 for rest of `staking.Rewards` **events**

[the remark this answers is not in the document]
Comment (stwiname) [reply packages/node-core/src/indexer/dictionary.service.ts:213]: That makes more sense. Can you update the comment to be that example please.

**Comment by jiqiang90** [comment]
> I don't think we're gaining anything with this.
> 
> We're making dictionary queries and discarding the results. We can be better about setting the queryEndBlock [here](https://github.com/subquery/subql/blob/585ceb34ea721f7c83dec71bf9d82d0f0f98cfc8/packages/node/src/indexer/fetch.service.ts#L431) and consider the keys of [mappedDictionaryQueryEntries](https://github.com/subquery/subql/blob/585ceb34ea721f7c83dec71bf9d82d0f0f98cfc8/packages/node-core/src/indexer/dictionary.service.ts#L138) to decide on the end value.

Here is the scenarios:

We have two datasources 
```yaml
  - kind: substrate/Runtime
    startBlock: 100
    mapping:
      handlers:
        - handler: handleStaking
          kind: substrate/EventHandler
          filter:
            module: staking
            method: Rewards

  - kind: substrate/Runtime
    startBlock: 50000
    mapping:
      handlers:
        - handler: handleStaking
          kind: substrate/CallHander
          filter:
            module: balances            # assume this extrinsic could first exist at block 20000
            method: Transfer
```

**Current implementations:**

We build two dictionaryQueryEntries, something like below:

- **entry100**:  {100: staking.Rewards}
- **entry50000**: {50000: staking.Rewards && balances.Transfer}

with const DICTIONARY_MAX_QUERY_SIZE = 10000;  

At beginning we take `entry100` and make dictionary queries, assume the queryEndBlock increase 100 -> 100100.. -> .. -> 40050

It is safe while query between block 100 to 40050, but when startBlock reach 40050 and `queryEndBlock` increased to 50050, we will update to use `entry50000`.

Since then, without consider staking.Reward query result,  dictionary query extrinsic `balances.Transfer` could return some blocks in between 40050 to 50050. 

Example, each responded batch [40060, 40070, 40071 ... 400200], [400300, 400500, 40900..].... [49500, 49700, 49999, 50010, 50020]

Because we set startBlock is 50000, so any blocks between 40050- 50000 should be to discard, in above example we should only keep 50010 and 50020

I understood we have filter in handler to discard these extrinsics. However these blocks still been queued and fetched, it could be a waste here. 

- Depend on extrinsics/event density, It might take a few more dictionary queries to get from 40050 to 50000. Worth case it could send (10000/ batch size) times extra queries with query distinct, could be worth without distinct. 
- With additional query entity , I checked it will be executed as an additional SQL on host db, it could slow down dictionary response
- More importantly, It these extra blocks will be fetched, slows down indexing. 

__________________________________________________________________________________________________

**In this PR solution:**

We not update the current `dictionaryQueryEntry` until the last block of dictionary query result tell us it beyond another dictionaryQueryEntry's startHeight. 

Use above example, we are keep use `entry100` and only query  `staking.Rewards` event.  

And until a dictionary returned [batch](https://github.com/subquery/subql/blob/c2f786eff351e66ad70dbfb01675223f73164cc4/packages/node-core/src/indexer/dictionary.service.ts#L216) like [49500,49800, 49900, 50200,50500], we knew the last block 50500 is beyond another entry `entry50000` startHeight, then we will discard the last query result, update current dictionaryQueryEntry to `entry50000`, and re-query again.

If dictionary return an empty array, which means there is no matching blocks until the `queryEndBlock`,  we can use the it as the last block to select current dictionaryQueryEntry 

In this case It should significantly reduce unnecessary queries.
